packages/devtools_app/lib/src/debugger/controls.dart

```diff
@@ -3,114 +3,138 @@
 // found in the LICENSE file.
 
 import 'package:flutter/material.dart' hide Stack;
+import 'package:provider/provider.dart';
 import 'package:vm_service/vm_service.dart';
 
+import '../auto_dispose_mixin.dart';
 import '../common_widgets.dart';
 import '../theme.dart';
 import '../ui/label.dart';
 import 'debugger_controller.dart';
 import 'scripts.dart';
 
-class DebuggingControls extends StatelessWidget {
-  const DebuggingControls({Key key, @required this.controller})
-      : super(key: key);
+class DebuggingControls extends StatefulWidget {
+  const DebuggingControls({Key key}) : super(key: key);
 
-  final DebuggerController controller;
+  @override
+  _DebuggingControlsState createState() => _DebuggingControlsState();
+}
+
+class _DebuggingControlsState extends State<DebuggingControls>
+    with AutoDisposeMixin {
+  DebuggerController controller;
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    controller = Provider.of<DebuggerController>(context);
+    addAutoDisposeListener(controller.isPaused);
+    addAutoDisposeListener(controller.resuming);
+    addAutoDisposeListener(controller.stackFramesWithLocation);
+  }
 
   @override
   Widget build(BuildContext context) {
+    final isPaused = controller.isPaused.value;
+    final resuming = controller.resuming.value;
+    final hasStackFrames = controller.stackFramesWithLocation.value.isNotEmpty;
+    final canStep = isPaused && !resuming && hasStackFrames;
```

Comment:
Why do you need stack frames to be able to step?

Reply:
before we were checking for `hasFrames`, but the state wasn't getting propagated to the buttons, so I left the requirement of needing frames. Maybe you don't need frames at all to be able to step? that would probably be a question for you.

Looking at the code, the step calls call the resume method on the vm service and just pass an isolate id as well as a StepOption, so maybe frames aren't needed. In what scenario would we have 0 stack frames while paused? and in that case if there is not a top stack frame, what code is there to step through?

Reply:
Yeah, from memory, there are times you can be paused but don't have any frames. You still want to be able to resume execution in those cases.

Reply:
> Yeah, from memory, there are times you can be paused but don't have any frames. You still want to be able to resume execution in those cases.

To be clear, Resume would still be enabled in this case - as the code is, though, the stepping buttons would not. Do we want to be able to step in this case as well as resume?

Reply:
Ah, no. Just having resume enabled would be sufficient. If we don't have an actual location - at least one frame - then stepping doesn't make sense.

Reply:
Okay - in that case, this code achieves that
- when paused with a top frame, resume and step buttons are enabled
- when paused without a top frame, resume is enabled but step buttons are disabled

Reply:
You can pause your application at any time so on the web side you can be paused in some random JS / SDK function. We won't show frames if paused in one of these locations but nothing prevents stepping from working. For the web at least, we continue stepping until you get to the next known Dart location.

I can see the argument to not support stepping in these scenarios however.

Reply:
I think this CL is fine to leave as is for now, as it mimics the previous intended behavior and handling the paused with 0 frames case seems like a larger issue we should think through. Filed https://github.com/flutter/devtools/issues/2687 to discuss and identify steps forward.